### PR TITLE
Migrate GitLab, BitBucket, GitHub interceptors to new interface

### PR DESCRIPTION
**Pull request by dibyom**
# Changes

This commit migrates the GitHub, GitLab, BitBucket interceptors to use the new
InterceptorRequest/Response types. Since these interceptors do not modify the
event body, this change is not a breaking change.

In addition, this commit also refactors the tests for the interceptors by
splitting valid and invalid scenarios into separate test cases.

Part of #271 and #778

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: pkg/interceptors/cel/cel.go [hunk 1]
```diff
@@ -124,23 +124,12 @@ func makeEvalContext(body []byte, h http.Header, url string) (map[string]interfa
...
 func (w *Interceptor) Process(ctx context.Context, r *triggersv1.InterceptorRequest) *triggersv1.InterceptorResponse {
-	b, err := json.Marshal(r.InterceptorParams)
-	if err != nil {
-		return &triggersv1.InterceptorResponse{
-			Continue: false,
-			Status:   status.Newf(codes.InvalidArgument, "failed to marshal json: %v", err),
-		}
+	p := triggersv1.CELInterceptor{}
+	if err := interceptors.UnmarshalParams(r.InterceptorParams, &p); err != nil {
+		return interceptors.Fail(status.Newf(codes.InvalidArgument, "failed to parse interceptor params: %v", err))
```
Comment (wlynch) [inline pkg/interceptors/cel/cel.go:129]: We should wrap the error (here and elsewhere in this PR) 
```suggestion
		return interceptors.Fail(status.Newf(codes.InvalidArgument, "failed to parse interceptor params: %w", err))
```

File: pkg/interceptors/github/github.go [hunk 2]
```diff
@@ -17,85 +17,88 @@ limitations under the License.
...
-			return nil, errors.New("no X-Hub-Signature header set")
+			return interceptors.Fail(status.New(codes.FailedPrecondition, "no X-Hub-Signature header set"))
 		}
-		secretToken, err := interceptors.GetSecretToken(request, w.KubeClientSet, w.GitHub.SecretRef, w.TriggerNamespace)
+
+		ns, _ := triggersv1.ParseTriggerID(r.Context.TriggerID)
+		secret, err := w.KubeClientSet.CoreV1().Secrets(ns).Get(ctx, p.SecretRef.SecretName, metav1.GetOptions{})
 		if err != nil {
-			return nil, err
+			return interceptors.Fail(status.New(codes.FailedPrecondition, fmt.Sprintf("error getting secret: %v", err)))
```
Comment (wlynch) [inline pkg/interceptors/github/github.go:77]:
```suggestion
			return interceptors.Fail(status.Newf(codes.FailedPrecondition,"error getting secret: %w", err))
```

**Review by wlynch — COMMENTED** [review]
Looks good! Few small changes.

File: pkg/interceptors/interceptors.go [hunk 3]
```diff
@@ -118,3 +122,34 @@ func GetInterceptorParams(i *triggersv1.EventInterceptor) map[string]interface{}
...
+		Continue: false,
+		Status:   status,
+	}
+}
+
+// Canonical updates the map keys to use the Canonical name
+func Canonical(h map[string][]string) http.Header {
+	c := map[string][]string{}
+	for k, v := range h {
+		c[http.CanonicalHeaderKey(k)] = v
```
[the remark this answers is not in the document]
Comment (dibyom) [reply pkg/interceptors/interceptors.go:138]: The input map in this case is already a map[string][]string type (so there should not be any duplicate keys). The only reason we need this is in theory someone could send in a non-canonical version of the header keys

File: pkg/interceptors/cel/cel.go (hunk 1 above)
Comment (dibyom) [reply pkg/interceptors/cel/cel.go:129]: That fails govet with `pkg/interceptors/github/github.go:65:28: Newf call has error-wrapping directive %w`

File: pkg/interceptors/github/github.go (hunk 2 above)
Comment (dibyom) [reply pkg/interceptors/github/github.go:77]: done.

File: test/signature.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+	"crypto/sha1" //nolint
+	"encoding/hex"
+	"fmt"
+)
+
+// HMACHeader generates a X-Hub-Signature header given a secret token and the request body
+// See https://developer.github.com/webhooks/securing/#validating-payloads-from-github
+func HMACHeader(secret string, body []byte) string {
+	h := hmac.New(sha1.New, []byte(secret))
+	_, _ = h.Write(body)
```
[the remark this answers is not in the document]
Comment (dibyom) [reply test/signature.go:14]: done.

**Review by wlynch — APPROVED** [review]
/lgtm

**Pull request merged**